### PR TITLE
H2 cash selection needs to handle multiple onlyFromIssuerParties and/…

### PR DESCRIPTION
**Pull request by cxyzhang0**
…or withIssuerRefs

Current CashSelectionH2Impl works only if set onlyFromIsuerParties contains 0 or one member.  Looks like with two members, the IN clause becomes ('A, B') instead of ('A', 'B').  The latter is our desired outcome.  
This PR fixes that, and the potential issue with withIssuerRefs as well by explicitly declaring and setting a query parameter for each member in the sets onlyFromIssuerParties and withIssuerRefs. That should be ok since I don't anticipate either set being large - as a matter of fact, Cash masks withIssuerRefs as an empty set as this moment.
I manually tested the onlyFromIssuerParties and withIssuerRefs.

# PR Checklist:

- [ ] Have you run the unit, integration and smoke tests as described here? https://docs.corda.net/head/testing.html
The current suit of tests should cover that.  I added one unit test to cover the multiple onlyFromParties. Manual testing of multiple members in either set has been done.
- [ ] If you added/changed public APIs, did you write/update the JavaDocs?
N/A
- [ ] If the changes are of interest to application developers, have you added them to the changelog, and potentially release notes?
N/A
- [ ] If you are contributing for the first time, please read the agreement in CONTRIBUTING.md now and add to this Pull Request that you agree to it.
Done.
Thanks for your code, it's appreciated! :)

### CODE REVIEW

File: finance/src/test/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2ImplTest.kt [hunk 1]
```diff
@@ -37,6 +37,21 @@ class CashSelectionH2ImplTest {
...
+        // Issuing Integer.MAX_VALUE will not cause an exception since PersistentCashState.pennies is a long
+        nCopies(2, Integer.MAX_VALUE).map { issueAmount ->
+            node.startFlow(CashIssueFlow(issueAmount.POUNDS, OpaqueBytes.of(1), mockNet.defaultNotaryIdentity))
+        }.transpose().getOrThrow()
+        // The spend must be more than the size of a single cash state to force the accumulator onto the second state.
+        // Also, here we test issuerConstraint with multiple onlyFromParties.
+        val request = CashPaymentFlow.PaymentRequest((Integer.MAX_VALUE + 1L).POUNDS, node.info.legalIdentities[0], true, setOf(node.info.legalIdentities[0], mockNet.defaultNotaryIdentity))
+        node.startFlow(CashPaymentFlow(request)).getOrThrow()
+    }
+
```
Comment (szymonsztuka) [inline finance/src/test/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2ImplTest.kt]: Maybe simplify separate simple this test and focus only on multiple issuers in where condition:
    `fun multiple issuers in issuerConstraint condition () {

        val node = mockNet.createNode()

        node.startFlow(CashIssueFlow(1.POUNDS, OpaqueBytes.of(1), mockNet.defaultNotaryIdentity)).getOrThrow()

        val request = CashPaymentFlow.PaymentRequest(1.POUNDS, node.info.legalIdentities[0], true, setOf(node.info.legalIdentities[0], mockNet.defaultNotaryIdentity))

        val paymentResult = node.startFlow(CashPaymentFlow(request)).getOrThrow()

        assertNotNull(paymentResult.recipient)
    }`

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt [hunk 2]
```diff
@@ -32,22 +32,40 @@ class CashSelectionH2Impl : AbstractCashSelection() {
...
+              AND vs.notary_name = ?
+            """)
+        if (onlyFromIssuerParties.isNotEmpty()) {
+            sb.append("""
+              AND ccs.issuer_key_hash IN (
+            """)
+            sb.append("?,".repeat(onlyFromIssuerParties.size))
+            // delete the last ","
+            sb.deleteCharAt(sb.length - 1)
+            sb.append(")")
```
Comment (szymonsztuka) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: What about (and as well for withIssuerRefs) (ignore my formatting):

` val repeats = generateSequence { "?" }
                            .take(onlyFromIssuerParties.size)
                            .joinToString (",")
                    " AND ccs.issuer_key_hash IN ($repeats)"
`

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt [hunk 3]
```diff
@@ -32,22 +32,40 @@ class CashSelectionH2Impl : AbstractCashSelection() {
     //       3) H2 does not support JOIN's in FOR UPDATE (hence we are forced to execute 2 queries)
     override fun executeQuery(connection: Connection, amount: Amount<Currency>, lockId: UUID, notary: Party?, onlyFromIssuerParties: Set<AbstractParty>, withIssuerRefs: Set<OpaqueBytes>, withResultSet: (ResultSet) -> Boolean): Boolean {
         connection.createStatement().use { it.execute("CALL SET(@t, CAST(0 AS BIGINT));") }
+        val sb = StringBuilder()
```
Comment (szymonsztuka) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: I would retain `+` concatenation, JVM should be able change it to`StringBuilder` anyway, eventually maybe a separate PR for this.

**Comment by cxyzhang0** [comment]
For some reason, the latest git push brought many unrelated commits.  So close this one and have opened #3659 as its replacement.

**Pull request closed without merging**